src/StateNode.ts

```diff
@@ -500,88 +510,56 @@ class StateNode<
-            toActionObject(action)
-          )
-        };
-      });
-    }
-
-    const allDelayedTransitions = flatten<
-      DelayedTransitionDefinition<TContext, TEvent>
-    >(
-      keys(afterConfig).map(delayKey => {
-        const delayedTransition = (afterConfig as Record<
-          string,
-          | TransitionConfig<TContext, TEvent>
-          | Array<TransitionConfig<TContext, TEvent>>
-        >)[delayKey];
-
-        const delay = isNaN(+delayKey) ? delayKey : +delayKey;
-        const eventType = after(delay, this.id);
-
-        this.onEntry.push(send<TContext, TEvent>(eventType, { delay }));
-        this.onExit.push(cancel(eventType));
+    const delayedTransitions = isArray(afterConfig)
+      ? afterConfig
+      : flatten(
+          keys(afterConfig).map(delay => {
+            const configTransition = afterConfig[delay];
+            const resolvedTransition = isString(configTransition)
+              ? { target: this.resolveTarget(configTransition) }
+              : configTransition;
+
+            return toArray(resolvedTransition).map(transition => ({
+              ...transition,
+              delay
+            }));
+          })
+        );
 
-        if (isString(delayedTransition)) {
-          return [
-            {
-              source: this,
-              target: this.resolveTarget(delayedTransition),
-              delay,
-              eventType,
-              actions: []
-            }
-          ];
-        }
+    return delayedTransitions.map((delayedTransition, i) => {
+      let { delay, target } = delayedTransition;
+      let delayRef: string | number;
+
+      if (isFunction(delay)) {
+        delayRef = `${this.id}:delay[${i}]`;
+        // TODO: this really doesn't seem to be used, consider removing
+        this.options.delays[delayRef] = delay; // TODO: util function
+      } else if (!isNaN(+delay)) {
+        delayRef = +delay;
+        delay = +delay;
+      } else {
+        delayRef = delay;
+      }
 
-        const delayedTransitions = toArray(delayedTransition);
-
-        return delayedTransitions.map(transition => ({
-          eventType,
-          delay,
-          ...transition,
-          source: this,
-          target:
-            transition.target === undefined
-              ? transition.target
-              : this.resolveTarget(transition.target),
-          cond: toGuard(transition.cond, guards),
-          actions: toArray(transition.actions).map(action =>
-            toActionObject(action)
-          )
-        }));
-      })
-    );
+      const eventType = after(delayRef, this.id);
 
-    allDelayedTransitions.sort((a, b) =>
-      isString(a) || isString(b) ? 0 : +a.delay - +b.delay
-    );
+      // TODO: check if using delayRef here (instead of delay) has fixed a bug
+      // check why we can't use delayRef here, how resolving of this works
+      // seems a little bit scattered across the codebase and inconsistent
```

Comment:
We can use `delayRef` I believe, it should work the same.

Reply:
It was what I did at first 😅 unfortunately it didn't work because we can't later (right now) retrieve original function for "inline" function delays covered by those lines:
https://github.com/davidkpiano/xstate/blob/8621b9db28bbb2ccc979b7775c1cea71ac5e022a/src/StateNode.ts#L533-L536
The entry in delays map is created by mutating `this.options.delays` (in kinda "target" state) **BUT** resolveSend gets called from "origin" (for a transition) state here: 
https://github.com/davidkpiano/xstate/blob/8621b9db28bbb2ccc979b7775c1cea71ac5e022a/src/StateNode.ts#L1160-L1164
and each~ state has its own copy of options, because they are not correctly forwarded i.e. here:
https://github.com/davidkpiano/xstate/blob/8621b9db28bbb2ccc979b7775c1cea71ac5e022a/src/StateNode.ts#L300

This might be a bug on master, gonna investigate how it behaves to see why this has slipped unnoticed. However - I'm unsure if mutating options is a good strategy for this anyway. I wouldn't expect those to be inserted there.

Reply:
After doing a quick investigation - luckily there is no bug on master but the implementatiin works sort of by accident in comparison to other similar stuff. Dont have time right now to describe in detail - gonna do that later.